uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/API/Query/LeaveRequestSelect.php

```diff
@@ -42,15 +45,81 @@ public function __construct($params) {
    * of the Public Holiday type will be returned.
    */
   private function addJoins() {
+    $this->addDateAndBalanceChangeJoins();
+    $this->addContractJoins();
+  }
+
+  /**
+   * Executes the query
+   *
+   * @return array
+   */
+  public function run() {
+    return $this->query->run();
+  }
+
+  /**
+   * Joins the Leave Request with Job Contracts, in order to only return
+   * the requests which overlap contracts. That is, the request where the start
+   * and/or the end date is between the start and end dates of of the contracts
+   * of the Leave Request contact.
+   */
+  private function addContractJoins() {
+    $this->query->join(
+      'INNER',
+      HRJobContract::getTableName(),
+      'jc',
+      [
+        'a.contact_id = jc.contact_id',
+        'jc.deleted = 0'
+      ]
+    );
+    $this->query->join(
+      'INNER',
+      HRJobContractRevision::getTableName(),
+      'jcr',
+      [
+        'jcr.id = (SELECT id
+                    FROM ' . HRJobContractRevision::getTableName() . ' jcr2
+                    WHERE
+                    jcr2.jobcontract_id = jc.id
+                    ORDER BY jcr2.effective_date DESC
+                    LIMIT 1)'
+      ]
+    );
```

Comment:
I'm not sure, but that could work. The problem is that the data structure of hrjobcontract is quite complex and it might not be so simple to change this. Also, the extension almost has no tests, so there's easy way to see if we'll break anything with such a change.
HRJobContract will be revisited/refactored some day and luckily we won't need those sub queries anymore. For now, we have to live with it :(